### PR TITLE
tidb-server 2.1.7 (new formula)

### PR DESCRIPTION
**Pull request by morgo**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?

Yes.  There is one major difference from guidelines, which is that I built from the git repo instead of a tarball.  I know the packaging guidelines prefer to build from a tarball, but when doing this with TiDB it will not know its version, since it relies on git tags.  i.e. commands like `select version(), select tidb_version()` will not behave as expected.

- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This provides tidb-server without TiKV or PD (the usual [required components](https://pingcap.com/docs/architecture/) for a distributed TiDB server).

It is useful for:
- Developers who wish to test TiDB's mysql compatibility locally
- Developers who want a MySQL-like system with a smaller footprint (the minimal install is much smaller than MySQL, MariaDB, Percona Server in terms of memory and disk space).

Notes:
- It requires go 1.11 to compile, but it looks like this version is skipped in homebrew packaging. 
 Dependencies install via go module support.
- I assume that a homebrew admin adds bottle support, so I have left this out for now.

Thank you in advance for your review!

### CODE REVIEW

**Comment by SMillerDev** [comment]
Thanks for the pull request, it mostly looks good to me. It seems your datadir definition is only used once so I think it's better to inline it.
Furthermore, your caveats are a lot longer than they would usually be. I think it'd be fine with just the note about slow benchmarks.

File: Formula/tidb-server.rb
Comment (fxcoudert) [inline Formula/tidb-server.rb]: I don't think we want that. Homebrew ships stable, non-debug configurations by default.

Comment (morgo) [reply Formula/tidb-server.rb]: To clarify: it is both stable, and non-debug.  My own estimate is that goleveldb is 5x slower than TiKV (uses RocksDB).

Why am I not using TiKV?  To use TiKV I would also need to install pd.  The [production configuration](https://pingcap.com/docs/architecture/) is also 3x pd and 3x tikv.  It is possible to do this on a local machine, but it's quite heavy.

Comment (fxcoudert) [reply Formula/tidb-server.rb]: We would want to ship this in the most common configuration, the one that upstream ships by default

**Comment by SMillerDev** [comment]
Good to hear, to me the connection issue sounds more like something that should be in the user manual/docs I'll wait for other maintainers' opinion on that.

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: This is the most common configuration, for a _single server_. The nuance comes because TiDB is a distributed system.  I looked at how this is done in the `cockroach` package (also a distributed database, but with a different architecture), and there is a [similar caveat](https://github.com/Homebrew/homebrew-core/blob/master/Formula/cockroach.rb#L33-L42).

**Comment by morgo** [comment]
> Good to hear, to me the connection issue sounds more like something that should be in the user manual/docs I'll wait for other maintainers' opinion on that.

For context, the mysql package also includes a [client connectivity](https://github.com/Homebrew/homebrew-core/blob/master/Formula/mysql.rb#L115-L116) caveat.  I also wish the mysql-client did not default to unix socket for "localhost", but since it is not distributed by TiDB, it is out of my control :(

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: For a longer answer to the question of _why the engine is different in a single server_:

TiDB is written in golang, and RocksDB is C++.  It is possible to link against it, but there would be FFI overhead.  It also complicates development, it's much easier to be a pure-go package.

In a distributed server environment, you can think of tidb as "sql processing nodes" and TiKV as "data nodes".  TiDB communicates to TiKV over gRPC. The TiKV server is the one doing the low-level data access, and is written in Rust instead of go.  For Rust to link against a C++ package there is no FFI overhead.

So to paraphrase, in the single server mode the TiDB server does double duty as sql processing and data storage.  It is actually the default (and what the tests-suite runs on), but it is not intended for production usage, since you lose the advantage of TiDB being a distributed system.

**Comment by morgo** [comment]
I have changed it so that it no longer requires a configuration file, and thus no longer requires a `-config` argument when running `tidb-server`.  My motivation for doing this, is so linuxbrew users can just type "tidb-server" without a config file argument, since linuxbrew does not yet [support services](https://github.com/Homebrew/brew/issues/5847).

Long term I think supporting a default config file is a better fix, so I've created issue https://github.com/pingcap/tidb/issues/9621 in TiDB.

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: @fxcoudert friendly followup ping :-)  Did my above comment address your concern?

File: Formula/tidb-server.rb
Comment (fxcoudert) [inline Formula/tidb-server.rb]: I'm a bit puzzled by this. If the project is designed to run with that datadir in a temporary directory, we shouldn't move it to a permanent directory. That completely changes the purpose.

(I won't comment on the practice of hardcoding paths in the source in the first place)

File: Formula/tidb-server.rb
Comment (fxcoudert) [inline Formula/tidb-server.rb]: What is the problem with the hashes? Has it been reported, so it gets fixed in the next release?

**Review by fxcoudert — DISMISSED** [review]

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: https://stackoverflow.com/questions/54133789/go-modules-checksum-mismatch

TiDB 2.1 upstream will compile with 1.11.2 for its lifetime, so the go.sum reflects how the now invalid hashes were generated.  There is no 1.11.2 in homebrew, which is fine, it will compile on the latest 1.12.x with the go.sum removed.  The upcoming 3.0 release will compile fine without this change, as it will use 1.12.x for its lifetime.

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: I am not sure of the historical reason it defaults to `/tmp/tidb`, it might be because of the test-suite.  If I leave it in `/tmp`, then the data is lost constantly.

The MySQL package uses a [cmake flag to set the datadir](https://github.com/Homebrew/homebrew-core/blob/master/Formula/mysql.rb#L56).  I also [commented [](here](https://github.com/Homebrew/homebrew-core/pull/37634#issuecomment-470970594)) why I made this change, and did not use a config file instead.  Long term this could be fixed by the `tidb-server` accepting a default config file.  I can not guarantee if it is fixed though, that it will be backported to the 2.1 branch.

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: @fxcoudert would you like me to link to this stackoverflow thread from the comment? The status of this PR is 'requested changes', but it is not straight forward to me how to make these changes.

File: Formula/tidb-server.rb
Comment (fxcoudert) [reply Formula/tidb-server.rb]: Then I'm tempted to say let's ship it as built and configured by upstream. They can fix it if they want, but we shouldn't have to work around their design decisions.

Comment (morgo) [reply Formula/tidb-server.rb]: I work for upstream if that was clear :-) I will work on the fix [described here](https://github.com/pingcap/tidb/issues/9621) shortly, but I can't guarantee a backport to the stable release series. It could break user systems if they have a file in `/etc/tidb.toml` and perform a minor version upgrade.  For the next major version, not a problem.

Shipping as configured though is not an expected behavior for users.

Comment (fxcoudert) [reply Formula/tidb-server.rb]: > Shipping as configured though is not an expected behavior for users

Puzzling. Is each distributor supposed to write its own configuration file, or modify the paths in the source? How are Debian and other linux distros doing?

File: Formula/tidb-server.rb
Comment (fxcoudert) [reply Formula/tidb-server.rb]: Right now I'm trying to understand what is the best way to handle things. Maybe for `go.sum` a clearer comment would be nice, but I don't understand why that's a problem for this specific formula and we never see it with other `go` formulas, even when `go` versions change.

Comment (morgo) [reply Formula/tidb-server.rb]: I assume that other formulas have either not adopted go modules, or have increased the minimum go that they support to avoid 1.11.2.  One could argue that TiDB should be more flexible with which version they pin a GA release to (1.11.x instead of 1.11.2).  But at the rate go changes, that is actually not that easy to do.  As the answerer commented in StackOverflow, unfortunately, breakage is common right now in Go modules (as it was only released in 1.11 and is not yet the default).

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: TiDB is not in Debian or other Linux distros.

I believe there are arch personal packages (I am not sure how that system works).  You can see how it works in [dbdeployer here](https://github.com/datacharmer/dbdeployer/blob/master/sandbox/tidb_templates.go#L44-L45).

To me this is the combination of two problems: the default, and me trying to make it work without a configuration file so it will work more elegantly on linuxbrew (which does not support services, so I want to make it work for starting without specifying a config file).

File: Formula/tidb-server.rb
Comment (morgo) [reply Formula/tidb-server.rb]: I have added a link to the StackOverflow thread in https://github.com/Homebrew/homebrew-core/commit/5d7eb8794d5b6756f985ae2ef5d9626a3eeaab75

**Comment by SMillerDev** [comment]
@fxcoudert force pushed, just tagging you so you can check again.

**Comment by morgo** [comment]
I have bumped the version to 2.1.7.

**Comment by morgo** [comment]
@fxcoudert Some time has elapsed, and it is not quite clear to me what changes are being requested. Do you mind providing additional context? Thanks!

**Comment by fxcoudert** [comment]
I'll let @SMillerDev handle the review. I don't particular like that we don't ship with default settings. If the project uses `/tmp` for data by default, I'd say let's leave it like that, and the user can surely configure it if they want. Otherwise, the project build system should use `prefix` as directory to store data.

**Comment by morgo** [comment]
Bump

**Comment by SMillerDev** [comment]
@morgo do your changes to tidb make the config dir configurable on build time? Or does it change the directory away from `/tmp`?

**Comment by morgo** [comment]
> @morgo do your changes to tidb make the config dir configurable on build time? Or does it change the directory away from `/tmp`?

They change the default away from `tmp`, but the dir is still configurable. Please see my comment here https://github.com/Homebrew/homebrew-core/pull/37634#discussion_r268410926

**Comment by SMillerDev** [comment]
In that case wouldn't it be better to have launchctl set those options?

**Comment by morgo** [comment]
> In that case wouldn't it be better to have launchctl set those options?

Please see my comment here https://github.com/Homebrew/homebrew-core/pull/37634#discussion_r268410926

Linuxbrew does not support services.

**Comment by SMillerDev** [comment]
It does not but you're not necessarily merging into linuxbrew here. A caveat would be good enough in my idea. The current solution seems a lot like overengineering.

**Comment by morgo** [comment]
> It does not but you're not necessarily merging into linuxbrew here. A caveat would be good enough in my idea. The current solution seems a lot like overengineering.

TiDB is a database. If the user omits the config file argument when starting the server, without this patch they will launch a server with data in their tmp directory. Even if it wasn't the tmp directory, having two potential locations is kind of like a localized split-brain scenario, and it can make it hard to reconcile the sources.

In either case, it looks like we are getting nowhere here. I am going to close this pull request, since it's not clear by which criteria we can proceed forward.

**Pull request closed without merging**
